design/design-doc-operations.md

```diff
@@ -0,0 +1,494 @@
+   FSI, then later omitting them from FSI. Operations will only run once.
+   They'll need to delete fields by explicitly setting them to `null`.
+
+An Operation's pipeline will run to completion once, as soon as you create it.
+I propose we support three ways - in addition to manual creation - to create an
+Operation:
+
+1. Compose an Operation using an XR
+1. Create an Operation on a regular schedule using a CronOperation
+1. Create an Operation when a resource changes using a WatchOperation
+
+Composing an Operation is hopefully self-explanatory. Operation doesn't have an
+abstraction resource (like an XR) by design - see [Alternatives
+Considered][#alternatives-considered] for more details.
+
+CronOperation and WatchOperation will be to Operation as CronJob is to Job in
+Kubernetes. They'll contain a template for an Operation, and create one as
+needed. Either on a cron schedule, or when a watched resource changes.
+
+Here's an example CronOperation:
+
+```yaml
+apiVersion: ops.crossplane.io/v1alpha1
+kind: CronOperation
+metadata:
+  name: cluster-rolling-upgrade
+spec:
+  # Crontab schedule
+  schedule: "0 12 * * *"
+  # How long after the scheduled time to wait before considering it too late to
+  # create the Operation.
+  startDeadline: 10m
+  # How many completed Operations to keep around.
+  successfulHistoryLimit: 3
+  failedHistoryLimit: 3
+  # Specifies how to treat concurrent executions of an operation - Allow
+  # (default), Forbid, or Replace.
+  concurrencyPolicy: Forbid
+  operationTemplate:
+    spec:
+      # How many times the pipeline can fail before the Operation fails.
+      retryLimit: 5
+      # A pipeline of functions - just like a Composition.
+      mode: Pipeline
+      pipeline:
+      - step: rolling-upgrade
+        functionRef:
+          name: function-rolling-upgrade
+        input:
+          targets:
+            apiVersion: example.org/v1
+            kind: KubernetesCluster
+            selector:
+              matchLabels:
+                ops.crossplane.io/eligible-for-rolling-update: "true"
+          batches:
+          - 0.01  # First upgrade 1% of clusters
+          - 0.1   # Then if that works, 10%
+          - 0.5   # Then 50%
+          - 1.0   # Then 100%
+          fromVersions:
+          - "v1.29"
+          toVersion: "v1.30"
+          versionField: spec.version
+          healthyConditions:
+          - Synced
+          - Ready
+status:
+  # Operations that're currently running
+  active:
+  - name: cluster-rolling-upgrade-amfka
+  lastScheduleTime: "2024-04-18T12:00:37+00:00"
+  lastSuccessfulTime: "2024-04-18T12:00:37+00:00"
+```
+
+This CronOperation makes a lot more sense than a bare Operation for the
+hypothetical rolling upgrade scenario. An Operation isn't long-running - it's
+akin to a single reconcile loop. So to upgrade a fleet of clusters in four
+batches you'd want the Operation to run (at least) four times, with each
+Operation handling the next largest batch.
```

Comment:
Will the `CronOperation` invoke the `Operation` job with any additional context about the time it was scheduled for, and something like the attempt number? I wonder if that would be helpful for building idempotent systems that derive their current state from the "current time". For example, instead of using `time.Now()`, you would use `extra.CronScheduleTime`

Reply:
No. I'm open to it but don't know how it'd be possible.

Does CronJob do this?

Reply:
Got it. We could easily add an annotation like that. The Operation's functions wouldn't see it though, so I'm not sure how useful it would be in practice.